finality-aleph/src/sync/task_queue.rs

```diff
@@ -0,0 +1,118 @@
+use std::{
+    cmp::Ordering,
+    collections::BinaryHeap,
+    fmt::{Debug, Formatter},
+};
+
+use log::warn;
+use tokio::time::{sleep, Duration, Instant};
+
+use crate::sync::LOG_TARGET;
+
+#[derive(Clone)]
+struct ScheduledTask<T> {
+    task: T,
+    scheduled_time: Instant,
+}
+
+impl<T> Eq for ScheduledTask<T> {}
+
+impl<T> PartialEq for ScheduledTask<T> {
+    fn eq(&self, other: &Self) -> bool {
+        other.scheduled_time.eq(&self.scheduled_time)
+    }
+}
+
+impl<T> PartialOrd for ScheduledTask<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<T> Ord for ScheduledTask<T> {
+    /// Compare tasks so that earlier times come first in a max-heap.
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.scheduled_time.cmp(&self.scheduled_time)
+    }
+}
+
+#[derive(Clone, Default)]
```

Comment:
I wouldn't derive `Clone` for it as it can mask uncorrect usage - I'm pretty sure we don't want to clone the queue? We do need to make it safe to use across different threads though.